### PR TITLE
First implementation of Triton inference engine

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
This is the implementation of the ai inference APIs for the Nvidia Triton Server.

Signed-off-by: Merlino [pierantonio.merlino@eurotech.com](mailto:pierantonio.merlino@eurotech.com)

### CODE REVIEW

**Comment by pierantoniomerlino** [comment]
Below are the screenshots for the dependencies taken from ClearyDefined.

Screenshot 2022-02-18 at 17 20 42
Screenshot 2022-02-18 at 17 21 30
Screenshot 2022-02-18 at 17 22 13
Screenshot 2022-02-18 at 17 22 39
Screenshot 2022-02-18 at 17 23 02
Screenshot 2022-02-18 at 17 23 25
Screenshot 2022-02-18 at 17 23 57
Screenshot 2022-02-18 at 17 25 30
Screenshot 2022-02-18 at 17 26 42
Screenshot 2022-02-18 at 17 27 48
Screenshot 2022-02-18 at 17 30 13
Screenshot 2022-02-21 at 11 19 41

**Comment by pierantoniomerlino** [comment]
Awaiting [CQ](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=23956)

File: kura/org.eclipse.kura.ai.triton.server/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ com.google.gson;version="2.7.0",
+ javax.annotation;version="1.3.5",
+ org.apache.commons.io;version="2.4.0",
+ org.eclipse.kura;version="[1.0,2.0)",
+ org.eclipse.kura.ai.inference;version="[1.0,2.0)",
+ org.eclipse.kura.configuration;version="[1.1,2.0)",
+ org.eclipse.kura.core.linux.executor;version="[1.0,2.0)",
+ org.eclipse.kura.core.util;version="[1.0,2.0)",
+ org.eclipse.kura.executor;version="[1.0,2.0)",
+ org.eclipse.kura.type;version="1.1.0",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/META-INF/MANIFEST.MF]: Please use range here

File: kura/org.eclipse.kura.ai.triton.server/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ lib/grpc-api-1.33.1.jar
+Bundle-ActivationPolicy: lazy
+Import-Package: com.google.common.base;version="25.0.0",
+ com.google.common.io;version="25.0.0",
+ com.google.common.util.concurrent;version="25.0.0",
+ com.google.gson;version="2.7.0",
+ javax.annotation;version="1.3.5",
+ org.apache.commons.io;version="2.4.0",
+ org.eclipse.kura;version="[1.0,2.0)",
+ org.eclipse.kura.ai.inference;version="[1.0,2.0)",
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/META-INF/MANIFEST.MF]: If your bundle implements an interface marked as `ProvideType` then the package that provides that interface should be inported with a range like `[current_major.current_minor, current_major.current_minor+1)`

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/org.eclipse.kura.ai.triton.server.TritonServerService.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+   Copyright (c) 2011, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/org.eclipse.kura.ai.triton.server.TritonServerService.xml]: Please updare copyright header

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerLocalManager.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
+package org.eclipse.kura.ai.triton.server;
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerLocalManager.java:13]: Missing copyright header

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+
+    private CommandExecutorService commandExecutorService;
+    private TritonServerServiceOptions options;
+    private TritonServerLocalManager tritonServerLocalManager;
+
+    private ManagedChannel grpcChannel;
+    private GRPCInferenceServiceBlockingStub grpcStub;
+
+    public void setCommandExecutorService(CommandExecutorService executorService) {
+        if (isNull(this.commandExecutorService)) {
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceImpl.java]: This check is probably not necessary

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/org.eclipse.kura.ai.triton.server.TritonServerService.xml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" activate="activate" configuration-policy="require" deactivate="deactivate" enabled="true" immediate="true" modified="updated" name="org.eclipse.kura.ai.triton.server.TritonServerService">
+   <implementation class="org.eclipse.kura.ai.triton.server.TritonServerServiceImpl"/>
+   <service>
+      <provide interface="org.eclipse.kura.configuration.ConfigurableComponent"/>
+      <provide interface="org.eclipse.kura.ai.inference.InferenceEngineService"/>
+   </service>
+   <property name="service.pid" type="String" value="org.eclipse.kura.ai.triton.server.TritonServerService"/>
+   <reference bind="setCommandExecutorService" cardinality="1..1" interface="org.eclipse.kura.executor.PrivilegedExecutorService" name="PrivilegedExecutorService" policy="static" unbind="unsetCommandExecutorService"/>
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/org.eclipse.kura.ai.triton.server.TritonServerService.xml]: If this reference is static, specifying the unbind method is probably not necessary

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+            this.commandExecutorService = executorService;
+        }
+    }
+
+    public void unsetCommandExecutorService(CommandExecutorService executorService) {
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceImpl.java]: If this reference is static, specifying the unbind method is probably not necessary

File: kura/org.eclipse.kura.ai.triton.server/META-INF/MANIFEST.MF (hunk 2 above)
Comment (salvatore-coppola) [reply kura/org.eclipse.kura.ai.triton.server/META-INF/MANIFEST.MF]: Replaced inference api import with `org.eclipse.kura.ai.inference;version="[1.0,1.1)"`

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
+package org.eclipse.kura.ai.triton.server;
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java:13]: Missing copyright header

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    private final int grpcPort;
+    private final int metricsPort;
+    private final boolean isLocal;
+
+    public TritonServerServiceOptions(final Map<String, Object> properties) {
+        requireNonNull(properties, "Properties cannot be null");
+        this.properties = Collections.unmodifiableMap(new HashMap<>(properties));
+
+        final Object propertyPorts = this.properties.get(PROPERTY_PORTS);
+        if (nonNull(propertyPorts) && propertyPorts instanceof Integer[]) {
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java]: If the `instanceof` expression returns true then `propertyPorts` is not null, there is no need for the `nonNull` check

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            Integer[] ports = (Integer[]) propertyPorts;
+            this.httpPort = ports[0];
+            this.grpcPort = ports[1];
+            this.metricsPort = ports[2];
```
Comment (nicolatimeus) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java:61]: Can the array elements be null?

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java (hunk 9 above)
Comment (nicolatimeus) [reply kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerServiceOptions.java]: The same applies to the checks below

File: kura/test/org.eclipse.kura.ai.triton.server.test/build.properties [hunk 11]
```diff
@@ -0,0 +1,21 @@
+#
+# Copyright (c) 2017, 2020 Eurotech and/or its affiliates and others
```
Comment (nicolatimeus) [inline kura/test/org.eclipse.kura.ai.triton.server.test/build.properties:2]: Update copyright header

File: kura/test/org.eclipse.kura.ai.triton.server.test/src/test/java/org/eclipse/kura/ai/triton/server/TritonServerServiceModelLoadTest.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
+package org.eclipse.kura.ai.triton.server;
```
Comment (nicolatimeus) [inline kura/test/org.eclipse.kura.ai.triton.server.test/src/test/java/org/eclipse/kura/ai/triton/server/TritonServerServiceModelLoadTest.java]: Missing copyright header

**Comment by MMaiero** [comment]
The CI reports the following error: "Missing requirement: org.eclipse.kura.ai.triton.server 1.0.0.qualifier requires 'java.package; javax.annotation 1.3.5' but it could not be found"

**Comment by nicolatimeus** [comment]
Changing the `javax.annotation` import version in provider bundle and test bundle to 1.2.0 should restore the build, we should also check if the bundle still works with the current develop

**Comment by salvatore-coppola** [comment]
> Changing the `javax.annotation` import version in provider bundle and test bundle to 1.2.0 should restore the build, we should also check if the bundle still works with the current develop

done!

**Comment by MMaiero** [comment]
Please rebase and solve the conflicts

File: kura/test/org.eclipse.kura.ai.triton.server.test/build.properties (hunk 11 above)
Comment (nicolatimeus) [reply kura/test/org.eclipse.kura.ai.triton.server.test/build.properties:2]: The header is still 2017, 2020

**Review by MMaiero — APPROVED** [review]

**Pull request merged**